text/XXX-unwind-FFI.md

```diff
@@ -0,0 +1,286 @@
+* Even with custom `panic` runtimes,
+  users may still inadvertently cause undefined behavior
+  by trying to link shared libraries that use different unwind runtimes.
+  For instance, there is no easy way to know whether a C shared library
+  on a non-Windows system
+  was compiled with the `-fexceptions` GCC or LLVM flag;
+  with this flag, such a library would be unwind-runtime-compatible
+  with a Rust shared library compiled with `-C panic.strategy=native`,
+  but without this flag, any attempt to unwind through the C stack frames
+  (regardless of the runtime used)
+  would be undefined behavior.
+  There is no way for Rust to guard against this scenario.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+This version of this RFC replaces a prior version
+that suggested a much more minimal change.
+Specifically, it suggested introduing a different function annotation,
+`#[unwind(Rust)]`, which would simply ensure that the marked function
+would not be marked `noexcept` and would not `abort` on `panic`.
+Because the current implementation of `libpanic_unwind`
+is compatible with native (C++ style) exceptions,
+no further guarantees were made
+regarding the behavior of the unwinding operation itself.
+
+Another possibility would be to add the `panic.runtime` options
+as new `strategy` values for the `-C panic=foo` flag.
+This may be simpler to teach and to implement,
+but crates using `-C panic=native`
+would be incompatible with all existing crates.
+
+The current proposal ensures
+that Rust provides full control over unwinding behavior
+for maximum compatibility with other languages' runtimes
+and with existing crates.
+The interaction between the `#[unwind(allowed)]` annotation
+and the `panic.strategy` option
+provides the maximum safety that the toolchain can provide
+without limiting Rust's ability to interoperate
+with shared libraries in other languages
+(particularly `libpng` and `libjpeg` as mentioned
+[above](#motivation)).
+
+# Prior art
+[prior-art]: #prior-art
+
+[RFC #1513](less-unwinding) introduced
+the `#![panic_runtime]` attributes, the `-C panic` compiler flag, and
+the corresponding `profile.foo.panic` Cargo option.
+
+The `#[unwind(allowed)]` annotation already exists as an unstable feature,
+though the current implementation does not impose any requirements
+on the `panic` implementation.
+
+As mentioned above, GCC and LLVM provide an `-fexceptions` flag
+that makes the C++ exception mechanism interoperable with C stackframes.
+The C standard does not refer to exceptions at all, however,
+and the C++ standard does not make any guarantees
+regarding the behavior of exceptions that escape from C++ stack frames
+into stack frames written in another language.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+* Should the `panic.runtime` option and its permitted values be renamed?
+* What is the minimal subset of features listed here that must be released
```

Comment:
@kornelski That's already provided by `#[unwind(allowed)]` in Nightly. I believe the reason it can't be stabilized yet is that the default `panic` runtime is not guaranteed to be stable, and `#[unwind(allowed)]` would (as explained above) expose this runtime code that may be compiled with a different runtime.

Regarding C++, unwinding in C _is_ C++ exception-propagating. That's why it requires the `-fexceptions` flag. C libraries compiled with GCC or LLVM without that flag _will not_ properly support unwinding.